### PR TITLE
Finished project

### PR DESCRIPTION
**Pull request by webmarkyn**

### CODE REVIEW

File: main.rb [hunk 1]
```diff
@@ -0,0 +1,102 @@
+# frozen_string_literal: true
+
+module Enumerable
+  def my_each
+    (0..length - 1).each do |i|
+      yield(self[i])
+    end
```
Comment (samgaco) [inline main.rb:9]: - [ ] my_each returns an Enumerator if no block is given

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+      temp << [proc.call(self[i])] if proc
+      temp << yield(self[i]) if block_given? && !proc
+    end
+    temp
+  end
+
+  def my_each_with_index
+    (0..length - 1).each do |i|
+      yield(self[i], i)
+    end
```
Comment (samgaco) [inline main.rb:27]: - [ ] my_each_with_index returns an enumerator if no block is given

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  end
+
+  def my_select
+    temp = []
+    (0..length - 1).each do |i|
+      temp << self[i] if yield(self[i])
```
Comment (samgaco) [inline main.rb:35]: - [ ] my_select returns an enumerator if no block is given

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      return true if block_given? && yield(item)
+      return true if item && !block_given?
+    end
+    false
+  end
+
+  def my_all?
+    each do |item|
+      if block_given?
+        return false unless yield(item)
```
Comment (samgaco) [inline main.rb]: - [ ] my_all? when a class is passed as an argument returns true if all of the collection is a member of such class
- [ ] my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [ ] my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    (0..length - 1).each do |i|
+      temp << self[i] if yield(self[i])
+    end
+    temp
+  end
+
+  def my_any?
+    each do |item|
+      return true if block_given? && yield(item)
+      return true if item && !block_given?
```
Comment (samgaco) [inline main.rb]: - [ ] my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class
- [ ] my_any? when a Regex is passed as an argument returns false if none of the collection matches the Regex
- [ ] my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

File: main.rb [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+      else return false unless item
+      end
+    end
+    true
+  end
+
+  def my_none?(condition = false)
+    each do |item|
+      if block_given?
+        return false if yield(item)
```
Comment (samgaco) [inline main.rb]: - [ ] my_none? when no block or argument is given returns true only if none of the collection members is true
- [ ] my_none? when a class is passed as an argument returns true if none of the collection is a member of such class
- [ ] my_none? when a Regex is passed as an argument returns true only if none of the collection matches the Regex

File: main.rb [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+      elsif item == condition then return false
+      end
+    end
+    true
+  end
+
+  def my_count(cond = true)
+    counter = 0
+    each do |item|
+      if block_given? && yield(item)
```
Comment (samgaco) [inline main.rb]: - [ ] my_count returns the number of items in enum through enumeration

File: main.rb [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+  def my_each
+    (0..length - 1).each do |i|
+      yield(self[i])
+    end
+  end
+
+  def my_map(&proc)
+    temp = []
+    (0..length - 1).each do |i|
+      temp << [proc.call(self[i])] if proc
```
Comment (samgaco) [inline main.rb]: - [ ] my_map returns a new array with the results of running block once for every element in enum.
- [ ] my_map returns an Enumerator if no block is given

File: main.rb [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+        counter += 1
+      end
+    end
+    counter
+  end
+
+  def my_inject(acc = 0)
+    each do |item|
+      acc = yield(acc, item)
+    end
```
Comment (samgaco) [inline main.rb:107]: - [ ] my_inject combines all elements of enum by applying a binary operation, specified by a block
- [ ] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method

**Review by samgaco — CHANGES_REQUESTED** [review]
Hi @webmarkyn :high_brightness: 

I leave some comments on what functionalities your methods are lacking. You can find a reference of how they are expected to work [here](https://ruby-doc.org/core-2.6.4/Enumerable.html).

Please, work on these required comments and submit them back for another code review once you are done.

------------------------
Moving on :arrow_down: 

* **[For your next code review requests]**  Add a short summary of changes in Pull Request description
-----------------------
* ###### STATUS OF THE PROJECT: REQUESTED CHANGES :stop_sign:

**Review by stratospherique — CHANGES_REQUESTED** [review]
> ### Review by Ahmed Mahfoudh TSE

### Great Progress @webmarkyn  :clap: . However, some changes needs to be made in order for us to approve your project. check below :arrow_down: 

* [x] enumerables my_each returns an Enumerator if no block is given
* [x] enumerables my_each_with_index returns an enumerator if no block is given
* [x] enumerables my_select returns an enumerator if no block is given
- `my_each, my_each_with_index, my_select` Should return an enumerator if no block is giving.
  - Self is not an Enumerator, check this [guide](https://blog.arkency.com/2014/01/ruby-to-enum-for-enumerator/) for mor info.

* [x] enumerables my_all? returns true if the block never returns false or nil
* [x] enumerables my_all? when no block or argument is given returns true when none of the collection members are false or nil
  - my_all? is returning false always.
* [x] enumerables my_map returns a new array with the results of running block once for every element in enum.
* [x] enumerables my_map returns an Enumerator if no block is given
  - Same thing as my_each
* [x] enumerables my_inject combines all elements of enum by applying a binary operation, specified by a block
  - You have an error in block's arguments proc & word. 
```
     argumentsNoMethodError:
       undefined method `length' for 0:Integer

```
* [x] enumerables my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method
  - A problem in the expected result.

### That's it :relieved: . Apply the required changes to your methods. When finished, kindly resubmit another code review form :pray:

File: main.rb [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+      each { |item| return false unless yield(item) }
+      return true
+    end
+    state = case param
+            when Regexp then check_regexp(false, param, true)
+            when Class then check_class(false, param, true)
+            else check(false, param, true)
+            end
+    state == false ? state : true
+  end
```
Comment (stratospherique) [inline main.rb:64]: * [ ] my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

File: main.rb [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+      each { |item| return true if yield(item) }
+      return false
+    end
+    state = case param
+            when Regexp then check_regexp(true, param)
+            when Class then check_class(true, param)
+            else check(true, param)
+            end
+    state == true ? state : false
+  end
```
Comment (stratospherique) [inline main.rb:51]: * [ ] my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

File: main.rb [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+      each { |item| return false if yield(item) }
+      return true
+    end
+    state = case param
+            when Regexp then check_regexp(false, param)
+            when Class then check_class(false, param)
+            else check(false, param)
+            end
+    state == false ? state : true
+  end
```
Comment (stratospherique) [inline main.rb:77]: * [ ] my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

File: main.rb [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+
+    return add_els(self) if args.include? :+
+
+    return divide_els(self) if args.include? :/
+
+    return subs_els(self) if args.include? :-
+
+    each { |item| acc = yield(acc, item) }
+    acc
+  end
```
Comment (stratospherique) [inline main.rb]: * [ ] my_inject combines all elements of enum by applying a binary operation, specified by a block
* [ ] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method

File: main.rb [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+
+  def my_map(&proc)
+    temp = []
+    (0..length - 1).each do |i|
+      temp << [proc.call(self[i])] if proc
+      temp << yield(self[i]) if block_given? && !proc
+    end
+    temp = to_enum if temp.empty?
+    temp
+  end
```
Comment (stratospherique) [inline main.rb]: * [ ] my_map returns a new array with the results of running block once for every element in enum.

Comment (stratospherique) [reply main.rb]: - Wrong method output

File: main.rb (hunk 13 above)
Comment (stratospherique) [reply main.rb]: - Take into consideration an array of strings
- Check method output in comparison to original method
an example ``Range.new(5,50).inject(2, :*)``

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Great work implementing the suggested requirements :clap: . However, after thoroughly testing your methods, some of your methods have other requirements to fulfill. check the code comments below :arrow_down: 

- **[REQUIRED]** delete uncessary comments from `main.rb`

### When finished, kindly resubmit another code review request :pray: 

Happy coding :v: 
**Ahmed Mahfoudh**

File: main.rb [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+
+    return add_els(self, acc) if args.include? :+
+
+    return divide_els(self, acc) if args.include? :/
+
+    return subs_els(self, acc) if args.include? :-
+
+    each { |item| acc = yield(acc, item) }
+    acc
+  end
```
Comment (Ispirett) [inline main.rb:123]: When running this 

 ` array = [1,2,3,4,5]`
`     array.my_inject { |sum, n| sum + n } == array.inject { |sum, n| sum + n }`
i get the following messages please address this issue.

```
  TypeError:
       no implicit conversion of Integer into String

```

**Review by Ispirett — CHANGES_REQUESTED** [review]
Hey @webmarkyn  great work :+1:  :100:  with changes and work done with this project.
There is just one bug  :bug:  see below.

Please use the correct git-flow  when submitting projects this project should use gitflow  

`feature-branch -> development `
**status: changes required
Isaac Browne**

**Review by maelfosso — APPROVED** [review]
# Congratulations :clap: You've done :smile: 

**The project is approved** :+1: You can go ahead and merge this branch.
Good luck with the next project.
The next time, Kindly implement the **Git flow** process.

Thanks
**Mael FOSSO**
[Github](https://github.com/maelfosso) | [Twitter](https://twitter.com/maelfosso)

**Pull request merged**